src/realm/string_interner.cpp

```diff
@@ -422,56 +417,52 @@ StringID StringInterner::intern(StringData sd)
     m_decompressed_strings.push_back({64, std::make_unique<std::string>(sd)});
     auto id = m_decompressed_strings.size();
     m_in_memory_strings.push_back(id);
-    add_to_hash_map(*m_hash_map.get(), h, id, 32);
-    size_t index = (size_t)m_top->get_as_ref_or_tagged(Pos_Size).get_as_int();
+    add_to_hash_map(m_hash_map, h, id, 32);
+    size_t index = (size_t)m_top.get_as_ref_or_tagged(Pos_Size).get_as_int();
     REALM_ASSERT_DEBUG(index == id - 1);
     bool need_long_string_node = c_str.size() >= 65536;
 
     // TODO: update_internal must set up m_current_long_string_node if it is in use
+    if (need_long_string_node && !m_current_long_string_node.is_attached()) {
 
-    if (need_long_string_node && !m_current_long_string_node) {
         if ((index & 0xFF) == 0) {
             // if we're starting on a new leaf, extend parent array for it
-            m_data->add(0);
+            m_data.add(0);
             m_compressed_leafs.push_back({});
-            m_current_long_string_node = std::make_unique<Array>(m_top->get_alloc());
-            m_current_long_string_node->set_parent(m_data.get(), m_data->size() - 1);
-            m_current_long_string_node->create(NodeHeader::type_HasRefs);
-            m_current_long_string_node->update_parent();
-            REALM_ASSERT_DEBUG(!m_current_string_leaf->is_attached() || m_current_string_leaf->size() == 0);
-            m_current_string_leaf->detach();
+            m_current_long_string_node.create(NodeHeader::type_HasRefs);
+            m_current_long_string_node.set_parent(&m_data, m_data.size() - 1);
+            m_current_long_string_node.update_parent();
```

Comment:
Still not sure why these 3 lines cannot be moved up before the "if". It is duplicated in the else part.

Reply:
`m_current_long_string_node.create(NodeHeader::type_HasRefs);` can
```
m_current_long_string_node.set_parent(&m_data, m_data.size() - 1);
 m_current_long_string_node.update_parent();
```
Cannot because we are resetting the parent, for the string node in the if/else and the parent is different based on which branch we are. Consequently, we are updating a different parent node with the ref.